src/intro-to-c/debugging/general-debugging.md

```diff
@@ -0,0 +1,293 @@
+
+# Example B: Bothersome Bananas.
+
+``` c
+#include <stdio.h>
+
+int main(void)
+{
+    int i;
+    char name[100] = "banana";
+
+    // This should print the name as a grid with 5 rows like:
+    // b a n a n a
+    // b a n a n a
+    // b a n a n a
+    // b a n a n a
+    // b a n a n a
+
+    for (i = 0; i < 5; i++)
+    {
+        for (i = 0; i < 100; i++)
+        {
+            printf("%c ", name[i]);
+        }
+        printf("\n");
+    }
+
+    return 0;
+}
+```
+
+``` {.terminal}
+b a n a n a
+```
+
+Hey, what gives? I wanted five bananas, not one!
+
+If we look a little closer, we'll notice that we're using `i` for both for loops. This means when the inner loop ends, `i` is at 100, causing the condition for the outer loop, `i < 5`, to evaluate to false and terminate the loop.
+
+If we want to see all five bananas, we need to use a separate variable for the inner loop. Here, we'll use `j` as the new counter.
+
+``` c
+#include <stdio.h>
+
+int main(void)
+{
+	int i, j;
+    char name[100] = "banana";
+
+    // This should print the name as a grid with 5 rows like:
+    // b a n a n a
+    // b a n a n a
+    // b a n a n a
+    // b a n a n a
+    // b a n a n a
+	
+    for (i = 0; i < 5; i++)
+    {
+        for (j = 0; j < 100; j++)
+        {
+            printf("%c ", name[j]);
+        }
+        printf("\n");
+    }
+	
+	return 0;
+}
+```
+
+If we run that, we get our expected result:
+
+``` {.terminal}
+b a n a n a
+b a n a n a
+b a n a n a
+b a n a n a
+b a n a n a
+```
+
+# Example C: Chaotic Change.
+
+``` c
+#include <stdio.h>
+
+int main(void)
+{
+    double change;
+    int dollars, quarters, dimes, nickels, pennies;
+
+    printf("Enter the amount of change: ");
+    scanf("%lf", &change);
+
+    while (change != 0)
+    {
+        if (change >= 1)
+        {
+            dollars++;
+            change -= 1;
+        }
+        else if (change >= 0.25)
+        {
+            quarters++;
+            change -= 0.25;
+        }
+        else if (change >= 0.10)
+        {
+            dimes++;
+            change -= 0.10;
+        }
+        else if (change >= 0.05)
+        {
+            nickels++;
+            change -= 0.05;
+        }
+        else if (change >= 0.01)
+        {
+            pennies++;
+            change -= 0.01;
+        }
+    }
+
+    printf("%d Dollars\n", dollars);
+    printf("%d Quarters\n", quarters);
+    printf("%d Dimes\n", dimes);
+    printf("%d Nickels\n", nickels);
+    printf("%d Pennies\n", pennies);
+
+    return 0;
+}
+```
+
+``` {.terminal}
+Enter the amount of change: 4.39
+
+```
+
+Why isn't my program printing anything?
```

Comment:
Why *isn't* this program printing anything? The first thing to notice is that the program does not finish, meaning there is some infinite loop in this program. Since there is only one loop here, it would be easy to deduce which loop the program is stuck in, but you may imagine placing print statements in every loop to see which one is the one causing problems.

My first hunch would be to print the value of `change` at the end of the while loop, and using this as a discussion point to learn something new about how C treats floating-point values. This is also a good time to mention there are multiple approaches to this problem, some don't use loops at all. A good point to bring up in debugging is to print way more decimal points than you normally would to expose weird edge cases, especially when it comes to floating-points.

This is just an example, but here are some of the print statements I may write to get started:
``` c
...
        else if (change >= 0.01)
        {
        		printf("pennies = %d\n", pennies);
        		printf("change = %0.2lf\n", change);
            pennies++;
            change -= 0.01;
        }
        printf("change = %0.20lf\n", change);
    }

    printf("%d Dollars\n", dollars);
    printf("%d Quarters\n", quarters);
    printf("%d Dimes\n", dimes);
...
```

Reply:
I used moving `change`'s decimal point and casting it to an integer as a solution in the update, but I'm wondering what other solutions would apply here?

Reply:
Another possible solution is to add a small "epsilon value," along the lines of 0.00000025, so that it is just enough to round up to the next cent. Personally however, the solution that I have seen to be the most reliable and predictable is to represent all money and operations in terms of an integer number of cents.

i.e.:

``` c
...
while (change != 0)
    {
        if (change >= 100)
        {
            dollars++;
            change = change - 100;
        }
        else if (change >= 25)
        {
            quarters++;
            change = change - 25;
        }
        else if (change >= 10)
        {
            dimes++;
            change = change - 10;
        }
...
```

A possible optimization from here is to also use modulus and integer division to perform all of the change counting in $O(1)$ time... but that probably goes beyond the scope of a simple intro to C tutorial... :)

``` c
...
dollars = change / 100;
change = change % 100;

quarters = change / 25;
change = change % 25;

dimes = change / 10;
change = change % 10;
...
```

Reply:
Thank you for the examples, and I hadn't considered the O(1) solution before.

I updated Example C to reflect the first solution you shared. It's essentially the same as before, but using actual values instead of expressions (e.g., `25` vs. `0.25 * factor`).

Reply:
Nice! I might suggest if you are going that route to make `change` itself into an integer. This would completely avoid all the possible issues with floating-point rounding. That's all I got!

Reply:
Thank you! I just made the change.